### PR TITLE
Faster matrix exponential computation

### PR DESCRIPTION
**Pull request by sylee957**
#### Brief description of what is fixed or changed

The previous implementation uses naive matrix power computation for matrix function series expansion, so it can be obvious that it can be slow if a jordan block is large

```python
from sympy import *
from sympy.matrices.sparsetools import banded
from pyinstrument import Profiler

l = Symbol('lambda')
m = banded(30, {0: l, 1: 1})
m = Matrix(m)

profiler = Profiler()
profiler.start()

m.exp()

profiler.stop()

print(profiler.output_text(unicode=True, color=True))
```

```
  _     ._   __/__   _ _  _  _ _/_   Recorded: 23:53:52  Samples:  12604
 /_//_/// /_\ / //_// / //_'/ //     Duration: 12.633    CPU time: 12.594
/   _/                      v3.0.3
```
It improves to 
```
  _     ._   __/__   _ _  _  _ _/_   Recorded: 23:53:24  Samples:  7308
 /_//_/// /_\ / //_// / //_'/ //     Duration: 7.320     CPU time: 7.281
/   _/                      v3.0.3
```
after using more direct formula, and using new `banded` constructor.

However, I see the `Matrix.exp` already takes about 7 seconds for computing jordan decomposition itself, even given a jordan block.
Which may mask the obvious speed disadvantage of using naive matrix powers, and large jordan cells may not appear that frequently for most of the real situations. But improvement is an improvement.

#### Release Notes

NO ENTRY

### CODE REVIEW

**Comment by czgdp1807** [comment]
Can we use the diaginalization of matrices? I used it in https://github.com/sympy/sympy/pull/17163/files#diff-78615da00c2c62bdf2a14971e2442e6aR711? I observed around 3.5 times improvement in performance by doing so. I will share the my empirical analysis here if there is a possibility of using the proposed idea.
Thanks.

**Comment by sylee957** [comment]
@czgdp1807 
Diagonalization is already incorporated in `jordan_form` I think
https://github.com/sympy/sympy/blob/a2d65f2a9c6c774404d602a941c09008e100025a/sympy/matrices/matrices.py#L1800

**Comment by oscarbenjamin** [comment]
@sylee957 Interesting. This definitely looks like an improvement.

Some time ago I implemented matrix exp for dsolve in #15449 (not merged) because I didn't realise it was already done for matrices (not sure if it's documented anywhere). I found though that:
1. I wanted the Jordan form of `exp(A*t)` and it was faster to compute this directly rather than compute exp (from the Jordan form of A) and then compute the Jordan form of that
2. Knowing that I wanted `exp(A*t)` where `A` has no symbols meant that I could compute the Jordan form more efficiently.
3. I wanted the realified version of the exponential that uses sin/cos where appropriate. I can rewrite as sin/cos after but the result came out in a simpler form if I did that directly for the Jordan blocks.

@czgdp1807 it looks like you are implicitly solving ODEs. It would probably be slower for now but your approach might generalise better to other stochastic processes if you frame them as ODEs and use dsolve.

File: sympy/matrices/matrices.py [hunk 1]
```diff
@@ -3130,6 +3130,47 @@ def dual(self):
...
+
+        .. [1] https://en.wikipedia.org/wiki/Matrix_function#Jordan_decomposition
+        """
+        size = self.rows
+        l = self[0, 0]
+        exp_l = exp(l)
+
+        bands = {0: exp_l}
+        for i in range(1, size):
+            bands[i] = exp_l / factorial(i)
```
Comment (oscarbenjamin) [inline sympy/matrices/matrices.py]: Is there a need to separate the i=0 case here?

**Comment by oscarbenjamin** [comment]
This looks good to me.

File: sympy/matrices/matrices.py [hunk 2]
```diff
@@ -3130,6 +3130,47 @@ def dual(self):
...
+
+        .. [1] https://en.wikipedia.org/wiki/Matrix_function#Jordan_decomposition
+        """
+        size = self.rows
+        l = self[0, 0]
+        exp_l = exp(l)
+
+        bands = dict()
+        for i in range(size):
+            bands[i] = exp_l / factorial(i)
```
Comment (oscarbenjamin) [inline sympy/matrices/matrices.py]: I guess this can just be a dict comprehension now:
```
bands = {i: exp_l / factorial(i) for i in range(size)}
```

**Comment by oscarbenjamin** [comment]
+1 to merge if you're done with this

**Comment by jksuom** [comment]
This looks fine.

**Pull request merged**